Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm

```diff
@@ -0,0 +1,358 @@
+                *outExceptionString = toErrorString(nil, contextsKey, @"'%@' is not a valid context", context);
+                return false;
+            }
+
+            parameters.contexts.value().add(typeMap.get().get(context));
+        }
+    }
+
+    if (NSArray *documentPatterns = properties[documentURLPatternsKey]) {
+        for (NSString *patternString in documentPatterns) {
+            auto pattern = WebExtensionMatchPattern::getOrCreate(patternString);
+            if (!pattern || !pattern->isSupported()) {
+                *outExceptionString = toErrorString(nil, documentURLPatternsKey, @"'%@' is not a valid pattern", patternString);
+                return false;
+            }
+        }
+
+        parameters.documentURLPatterns = makeVector<String>(documentPatterns);
+    }
+
+    if (NSArray *targetPatterns = properties[targetURLPatternsKey]) {
+        for (NSString *patternString in targetPatterns) {
+            auto pattern = WebExtensionMatchPattern::getOrCreate(patternString);
+
+            // Any valid pattern is allowed, not just supported schemes.
+            if (!pattern || !pattern->isValid()) {
+                *outExceptionString = toErrorString(nil, targetURLPatternsKey, @"'%@' is not a valid pattern", patternString);
+                return false;
+            }
+        }
+
+        parameters.targetURLPatterns = makeVector<String>(targetPatterns);
+    }
+
+    if (NSString *identifier = objectForKey<NSString>(properties, idKey, false)) {
+        if (!identifier.length) {
+            *outExceptionString = toErrorString(nil, idKey, @"it must not be empty");
+            return false;
+        }
+
+        parameters.identifier = identifier;
+    } else if (NSNumber *identifier = objectForKey<NSNumber>(properties, idKey))
+        parameters.identifier = identifier.stringValue;
+
+    if (NSString *parentIdentifier = objectForKey<NSString>(properties, parentIdKey, false)) {
+        if (!parentIdentifier.length) {
+            *outExceptionString = toErrorString(nil, parentIdKey, @"it must not be empty");
+            return false;
+        }
+
+        parameters.parentIdentifier = parentIdentifier;
+    } else if (NSNumber *parentIdentifier = objectForKey<NSNumber>(properties, parentIdKey))
+        parameters.parentIdentifier = parentIdentifier.stringValue;
+
+    if (NSString *title = properties[titleKey]) {
+        if (!title.length) {
+            *outExceptionString = toErrorString(nil, titleKey, @"it must not be empty");
+            return false;
+        }
+
+        if (parameters.type == WebExtensionMenuItemType::Separator) {
+            *outExceptionString = toErrorString(nil, titleKey, @"it cannot be provided when type is 'separator'");
+            return false;
+        }
+
+        parameters.title = title;
+    }
+
+    if (JSValue *clickCallback = properties[onclickKey]) {
```

Comment:
Nit: `onclickKey`.

Reply:
I meant it should be `onClickKey`, but that's not right because it's the `onclick` event.